Update tree view to include new metadata

marcellamaki opened the pull request:
## Summary
### Description of the change(s) you made

Fixed #3206 

Adds titles and metadata in the Resource Panel for Studio.
Does not add or update the Learning Activity Label or icons (follow up PR forthcoming)



### Manual verification steps performed
1.  Added test data with ids from le-utils to test the parsing of ids and strings in each of the new categories
### Screenshots (if applicable)

<img width="1051" alt="Screen Shot 2022-03-17 at 4 49 22 PM" src="https://user-images.githubusercontent.com/17235236/158892742-d8daa484-21c2-4e38-994e-becc5368d716.png">

___
## Reviewer guidance
### How can a reviewer test these changes?
<!-- If not applicable, please delete this section -->


### Are there any risky areas that deserve extra testing?
<!-- If not applicable, please delete this section -->


## References
<!--
Additional, helpful things to add in this section:
 * links to mockups or specs for new features
 * links to the diffs for any dependency updates, e.g. in iceqube or the perseus plugin
 * references to relevant issues or Notion projects
-->

## Comments
<!-- Additional comments may be added here -->

----

### Contributor's Checklist
<!-- After saving the PR, come through to tick off completed checklist items. Delete any sections that are not applicable to your PR -->

PR process:

- [ ] If this is an important user-facing change, PR or related issue the `CHANGELOG` label been added to this PR. Note: items with this label will be added to the [CHANGELOG](https://github.com/learningequality/studio/blob/master/CHANGELOG.md) at a later time
- [ ] If this includes an internal dependency change, a link to the diff is provided
- [ ] The `docs` label has been added if this introduces a change that needs to be updated in the [user docs](https://kolibri-studio.readthedocs.io/en/latest/index.html)?
- [ ] If any Python requirements have changed, the updated `requirements.txt` files also included in this PR
- [ ] Opportunities for using Google Analytics here are noted
- [ ] Migrations are [safe for a large db](https://www.braintreepayments.com/blog/safe-operations-for-high-volume-postgresql/)

Studio-specifc:

- [ ] All user-facing strings are translated properly
- [ ] The `notranslate` class been added to elements that shouldn't be translated by Google Chrome's automatic translation feature (e.g. icons, user-generated text)
- [ ] All UI components are LTR and RTL compliant
- [ ] Views are organized into `pages`, `components`, and `layouts` directories [as described in the docs](https://github.com/learningequality/studio/blob/vue-refactor/docs/architecture.md#where-does-the-frontend-code-live)
- [ ] Users' storage used is recalculated properly on any changes to main tree files
- [ ] If there new ways this uses user data that needs to be factored into our [Privacy Policy](https://github.com/learningequality/studio/tree/master/contentcuration/contentcuration/templates/policies/text), it has been noted.


Testing:

- [ ] Code is clean and well-commented
- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Any new interactions have been added to the [QA Sheet](https://docs.google.com/spreadsheets/d/1HF4Gy6rb_BLbZoNkZEWZonKFBqPyVEiQq4Ve6XgIYmQ/edit#gid=0)
- [ ] Critical and brittle code paths are covered by unit tests
___

### Reviewer's Checklist
#### This section is for reviewers to fill out.

- [ ] Automated test coverage is satisfactory
- [ ] PR is fully functional
- [ ] PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- [ ] External dependency files were updated if necessary (`yarn` and `pip`)
- [ ] Documentation is updated
- [ ] Contributor is in AUTHORS.md

marcellamaki commented:
@rtibbles @bjester -- I am going to make the PR description more robust for Jessica's review with screenshots to ensure I am displaying all scenarios correctly, but I would be happy to have some eyes on the code to see if I have actually done what I think I have done :) So, if you have time to take a look, that would be great. Thank you!

rtibbles reviewed (dismissed):
A couple of updates needed.

marcellamaki commented on contentcuration/contentcuration/frontend/channelEdit/components/ResourcePanel.vue:
```diff
@@ -554,6 +605,89 @@
         return this.node[field] || this.defaultText;
       },
+      matchIdToString(ids) {
+        return ids.map(i => this.translateMetadataString(camelCase(i))).join(', ');
```
So...translate each string in the array, and then pass the current language of the UI and the translated array into the list formatter?

rtibbles replied:
Yes.

marcellamaki replied:
After reviewing the figma, it looks like using a conjunction is not what is in the design (just commas)... I did try out adding the list format and it worked so I can go back and add this in quite easily, but I did not commit those changes because I'm not sure what the benefit would be over doing a join

rtibbles replied:
The benefit over doing a `join(', ')` is that it will internationalize the listing, depending on locale. Not every language uses commas to generate lists in the way that English does, and it will ensure that there are no gotchas for RTL languages.

Please also rename the function to something like `metadataListText` or somesuch to make it more descriptive!

marcellamaki commented:
### Updated views 

This PR has been updated to include the card and resource panel display of metadata on default, compact, and comfortable views with the new metadata icons. 

At this point, the only outstanding piece will be to add completion and duration display as a follow up issue

Figma, for reference 

|   |   |
|---|---|
| Compact view, with resource panel open (part 1) | <img width="1437" alt="Screen Shot 2022-06-27 at 10 27 04 AM" src="https://user-images.githubusercontent.com/17235236/175967568-1fe77586-7895-46b2-a9cb-9fd15f507256.png"> |
| Compact view, with resource panel open (part 2): previous step has multiple learning activities | <img width="1439" alt="Screen Shot 2022-06-27 at 10 27 13 AM" src="https://user-images.githubusercontent.com/17235236/175967620-74e5a3f9-6b81-4229-979f-c1521caf1a1e.png"> |
| Compact view, with resource panel open: current item has multiple learning activities | <img width="1434" alt="Screen Shot 2022-06-27 at 10 28 30 AM" src="https://user-images.githubusercontent.com/17235236/175967639-4c2c5b94-f940-494a-bbd3-3090f51b4aa5.png"> |
| Comfortable view (with description) | <img width="1428" alt="Screen Shot 2022-06-27 at 10 28 40 AM" src="https://user-images.githubusercontent.com/17235236/175967936-77accea4-5e52-441a-915b-2fa78bd8e680.png"> |
| Comfortable view (w/o description) | <img width="1435" alt="Screen Shot 2022-06-27 at 10 28 53 AM" src="https://user-images.githubusercontent.com/17235236/175968025-e1e65106-d2a5-4a6f-a76f-f063cbbb8e10.png"> |
| Default view  | <img width="1438" alt="Screen Shot 2022-06-27 at 10 29 03 AM" src="https://user-images.githubusercontent.com/17235236/175968261-5ca00f9d-50d8-4dbe-9ffe-c43c78922d80.png"> |

marcellamaki commented on contentcuration/contentcuration/frontend/channelEdit/components/ResourcePanel.vue:
```diff
@@ -139,7 +158,20 @@
+          <DetailsRow
+            v-if="!isTopic"
+            :label="translateMetadataString('duration')"
+            :text="duration(node.extra_fields)"
```
to be addressed in follow up

MisRob reviewed (changes requested):
Thank you, @marcellamaki . Below you can find notes that I think would be good to address (some may need only a clarification) before we merge. I’ll also post some comments in the code and mark those that are not blocking appropriately.

### 1. Possible visual regressions in other features

Updates in this PR affect other features, not only the channel tree navigation. One example would be the “Import from channels” feature:

| Before | After |
| --------- | ------ |
| ![before](https://user-images.githubusercontent.com/13509191/178005435-0dbaf239-d4bd-4446-b5f7-371709e0b44a.png) | ![after](https://user-images.githubusercontent.com/13509191/178005465-28b9a86e-b1a4-4653-b3c7-5d435ca127e6.png) |

Before we merge this, I’d appreciate clarification on whether this was intended because from the user's point of view, the new version seems to be confusing to me since now they won’t be able to understand resources presented very well when importing from other channels.

I haven’t checked other features that may be affected by the update, this is only one example. In general, I’d suggest uploading before/after changes for every UI update. I know it can be lots of work and you already made a great effort and uploaded tons of screenshots, but don’t know about any better way how to prevent regressions or communicate that there were changes in places beyond the new feature implemented. As for the reviewer role, it’s hard to find all the places to check, and even when one can derive something from code, it’s hard to recognize whether changes were intended or not.

### 2. @formatjs/intl

Is there any advantage of installing the whole library when compared to [using polyfills together with capability detection](https://formatjs.io/docs/polyfills/intl-listformat/#dynamic-import--capability-detection)? Generally speaking, I’d say it’s better to include only what’s needed so that the app won’t grow and that native APIs will be used whenever possible. I can see that @rtibbles  made some suggestions about polyfills earlier, any thoughts?

### 3. Lack of tests for new components or updates to existing ones

One of the examples would be the new `ContentNodeLearningActivityIcon` which contains lots of conditional logic which causes that (at least to me) it takes a while to understand all scenarios that could happen even when it’s a relatively simple component. There are also updates to existing components like `ResourcePanel` and `Thumbnail`. There’s no need to have a complex test suite, just a couple of main scenarios would be helpful.

MisRob commented on contentcuration/contentcuration/frontend/channelEdit/components/ContentNodeListItem/index.spec.js:
```diff
@@ -42,6 +42,11 @@ describe('ContentNodeListItem', () => {
         node: DOCUMENT_NODE,
       },
+      computed: {
+        hasMetadataToDisplay() {
```
This is looking to be an obsolete piece as there is no such computed in the `ContentNodeListItem`

MisRob commented on contentcuration/contentcuration/frontend/channelEdit/components/ResourcePanel.vue:
```diff
@@ -127,10 +125,21 @@
+            v-if="!isTopic"
+            :label="translateMetadataString('level')"
+            :text="level(node.grade_levels)"
+            notranslate
```
Rather for my understanding. I assume that categories, learning activities and grade levels are defined by constants and are not user-generated data so I can't understand very well why there's the need to use `notranslate` attribute, could you please elaborate?

MisRob commented on contentcuration/contentcuration/frontend/shared/views/files/Thumbnail.vue:
```diff
@@ -115,10 +89,6 @@
         required: false,
         default: false,
       },
-      isEmpty: {
```
Relates to (1) of the main comment, but haven't checked visually in this case. `isEmpty` property is removed here but from the code, I can see that it is still actively used by some features that import `files/Thumbnail.vue`, [for example here in the clipboard](https://github.com/learningequality/studio/blob/unstable/contentcuration/contentcuration/frontend/channelEdit/components/Clipboard/ContentNode.vue#L44). It'd be good to make sure that it can be really deprecated from other features as well (and if yes, to clean up all references to it).

MisRob commented on contentcuration/contentcuration/frontend/channelEdit/components/ContentNodeListItem/index.vue:
```diff
@@ -267,6 +304,26 @@
         }
       },
+      matchIdToString(ids) {
+        return ids.map(i => this.translateMetadataString(camelCase(i))).join(', ');
```
This resembles the logic in the `ResourcePanel` to me. If that's right, it'd suggest to use `Intl.FormatList` here too.

MisRob commented on contentcuration/contentcuration/frontend/shared/views/ContentNodeLearningActivityIcon.vue:
```diff
@@ -0,0 +1,124 @@
+    padding: 10px;
+    margin: 4px;
+    font-weight: bold;
+    background-color: #dedede;
```
KDS theme palette seems to be available in Studio, so I'd suggest using that instead of hardcoded colors. Same applies to the `.small-chip`'s `background-color` below.

MisRob commented on contentcuration/contentcuration/frontend/shared/views/ContentNodeLearningActivityIcon.vue:
```diff
@@ -0,0 +1,124 @@
+    mixins: [metadataTranslationMixin],
+
+    props: {
+      learningActivity: {
```
Not blocking, just would be nice. I think that `learningActivities` name for this prop would make a bit more sense because we can receive more than one. Also having a validator function would help to understand what values are expected here. The same applies to the same property in the `Thumbnail` component.

MisRob commented on contentcuration/contentcuration/frontend/channelEdit/components/ResourcePanel.vue:
```diff
@@ -3,13 +3,18 @@
+          <slot name="navigation"></slot>
+          <ContentNodeLearningActivityIcon v-if="isTopic" :isTopic="true" includeText chip />
+          <ContentNodeLearningActivityIcon
+            v-else-if="hasLearningActivities"
```
Not blocking, this can stay as is. Just an idea for rows 9-15 if you'd find it helpful. Maybe having another component, `TopicIcon`, separated from `ContentNodeLearningActivityIcon` would help with understanding and using `ContentNodeLearningActivityIcon`.

marcellamaki commented:
Hi @MisRob - thank you for your comprehensive review!
Following this and the discussion in slack, I've added back in the original thumbnail and created a new temporary component as we transition to the new thumbnail component to prevent regressions. I also implemented the check for the polyfill, and added a comment re: $formatList. In our i18n setup, we are using other formatting options such as $formatNumber and $formatDate. However, due to a bug in formatjs, $formatList was excluded, which is why we are having to do this polyfill in general. I added a comment to explain, and a link to the upgraded version of (it was just fixed last week while I was working on this! but is a huge version jump). 
I also added tests for ContentNodeLearningActivityIcon, which also prompted me to do some simplifying and refactoring in that file. 😄

MisRob commented:
Thank you, @marcellamaki. Regarding polyfills, thank you for the comment, I wasn't aware of the relationship to `vue-intl` and the bug. So now I'm wondering whether my first comment about not using the whole `@formatjs/intl` library was relevant because I assume we somehow use it under the hood, anyways? I'm sorry for the complications if that's the case. I'll leave the final decision on you and @rtibbles and try to follow to understand this better. Also, if the whole polyfill thing shows up to be complicated, I think it'd be fine not to use `Intl` in this PR and do it as a follow-up issue.

---

If we want to keep using just polyfills instead of bringing in the whole library as I suggested earlier and is implemented in your latest changes, some last touches would be needed:
- `Intl.ListFormat` polyfill has two dependencies (Intl.getCanonicalLocales or polyfill, Intl.Locale or polyfill). If I can understand their documentation ([here](https://formatjs.io/docs/polyfills/intl-listformat/#requirements) and [here](https://formatjs.io/docs/polyfills)), we need to load them too.
- I can see two `polyfill` functions, one in the `ResourcePanel` and another in `shared/i18n/index.js`. One of them needs to be removed. I think that keeping the latter makes more sense when compared to loading the polyfill in a component. Just note that the `polyfill` function needs to be called in that file.
- Some cleanup of the previous implementation is needed - uninstallation of `@formatjs/intl` and not using `createIntl`.

To put this all together, I think that we only need to have `@formatjs/intl-listformat` , `@formatjs/intl-getcanonicallocales`,`@formatjs/intl-locale` installed and the   `polyfill` function could look something like this (haven’t tried, just pulled pieces from documentation):
```
import {shouldPolyfill as shoudPolyfillCanonicalLocales } from '@formatjs/intl-getcanonicallocales/should-polyfill'
import {shouldPolyfill as shouldPolyfillLocale } from '@formatjs/intl-locale/should-polyfill'
import {shouldPolyfill as shouldPolyfillListFormat} from '@formatjs/intl-listformat/should-polyfill'

async function polyfill(locale) {
  if (shoudPolyfillCanonicalLocales()) {
    await import('@formatjs/intl-getcanonicallocales/polyfill')
  }
	if (shouldPolyfillLocale()) {
    await import('@formatjs/intl-locale/polyfill')
  }
	const unsupportedLocale = shouldPolyfillListFormat(locale)
  if (!unsupportedLocale) {
    return
  }
  await import('@formatjs/intl-listformat/polyfill-force')
  await import(`@formatjs/intl-listformat/locale-data/${unsupportedLocale}`)
}
```

MisRob commented:
New tests are looking good to me, thank you. I also skimmed briefly through `ImageOnlyThumbnail` and haven't noticed anything outstanding. I didn't study styles much though as I can see most of it comes from the original thumbnail component that have worked well for some time I think.

marcellamaki commented:
@MisRob, following the discussion in the planning meeting, I've updated just to remove the polyfill, and use the default Intl.ListFormat (since it aligns with the browsers we support). I hope after that final commit is reviewed, this is ready, but please let me know if there is anything else you see 😄

MisRob reviewed (approved):
Thanks for your patience, and sorry for the complications about polyfills, I misunderstood our needs and dependencies.

rtibbles commented:
> I misunderstood our needs and dependencies.

I think the lack of clarity wasn't helped by me either - plenty of blame to go around!

pcenov commented:
Looks good to me, have reported one related issue separately: https://github.com/learningequality/studio/issues/3664

marcellamaki commented:
Thanks @pcenov -- I will assign that issue to myself to fix the remaining problems!
